src/crimson/osd/osd.cc

```diff
@@ -1114,6 +1132,82 @@ seastar::future<> OSD::_handle_osd_map(Ref<MOSDMap> m)
   });
 }
 
+/*
+ * Compare between the previous last_map we had to
+ * each one of the added_maps.
+ * Track all of the changes relevant in pg_num_history.
+ */
+seastar::future<> OSD::track_pools_and_pg_num_changes(
+  const std::map<epoch_t, OSDMapService::local_cached_map_t>& added_maps,
+  ceph::os::Transaction& t)
+{
+  LOG_PREFIX(OSD::track_pools_and_pg_num_changes);
+  epoch_t first = added_maps.begin()->first;
+  epoch_t last = added_maps.rbegin()->first;
+
+  // Unless this is the first start of this OSD,
+  // lastmap should be the newest_map we have.
+  OSDMapService::local_cached_map_t lastmap;
+
+  if (superblock.maps.empty()) {
+    DEBUG(" no maps stored, this is probably the first start of this osd");
+    lastmap = added_maps.at(first);
+  } else {
+    if (first > superblock.get_newest_map() + 1) {
+      ceph_assert(first == superblock.cluster_osdmap_trim_lower_bound);
+      DEBUG(" can't get previous map {} first start of this osd after a map gap",
+	  superblock.get_newest_map());
+    }
+  }
+  return pg_shard_manager.get_local_map(
+    superblock.get_newest_map()
+  ).then([this, FNAME, &t, added_maps, last](auto lastmap) {
+    for (auto& [current_added_map_epoch, current_added_map] : added_maps) {
+      _track_pools_and_pg_num_changes(t, lastmap,
+                                    current_added_map,
+                                    current_added_map_epoch);
+      lastmap = current_added_map;
```

Comment:
I think that we might have an issue with `lastmap` here.
In classic, before calling `_track_pools_and_pg_num_changes` with each added_map, we try to obtain `lastmap`.
There are three possible options for `lastmap`:
1) "the newest_map we have" - most likely
2) "added_maps.at(first)" - first start of this osd
3) empty lastmap - first start after a map gap

Regarding the third option, each time we iterate over our added maps, we update `lastmap`. When we pass an empty map to `_track_pools_and_pg_num_changes`, there will be no information to work with and we will move on to the next (iteration) map in `added_maps`.

---

With the implementation here, `OSDMapService::local_cached_map_t lastmap;` will _never_ be used when iterating over `added_maps`. There is **another** `lastmap` instance which we got from `get_local_map` that is actually used while iterating.

Reply:
Thanks for catching that. I'll re-write and update it.